src/request.rs

```diff
@@ -265,19 +282,38 @@ impl<State> Request<State> {
     ///
     /// # Errors
     ///
-    /// Yields an `Err` if the parameter was found but failed to parse as an
+    /// Yields a `ParamError::ParsingError` if the parameter was found but failed to parse as an
     /// instance of type `T`.
     ///
-    /// # Panics
+    /// Yields a `ParamError::NotFound` if `key` is not a parameter for the route.
     ///
-    /// Panic if `key` is not a parameter for the route.
-    pub fn param<T: FromStr>(&self, key: &str) -> Result<T, T::Err> {
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # use async_std::task::block_on;
+    /// # fn main() -> Result<(), std::io::Error> { block_on(async {
+    /// #
+    /// use tide::{Request, Result};
+    ///
+    /// async fn greet(req: Request<()>) -> Result<String> {
+    ///     let name = req.param("name").unwrap_or("world".to_owned());
+    ///     Ok(format!("Hello, {}!", name))
+    /// }
+    ///
+    /// let mut app = tide::new();
+    /// app.at("/hello").get(greet);
+    /// app.at("/hello/:name").get(greet);
+    /// app.listen("127.0.0.1:8080").await?;
+    /// #
+    /// # Ok(()) })}
+    /// ```
+    pub fn param<T: FromStr>(&self, key: &str) -> Result<T, ParamError<T::Err>> {
```

Comment:
Isn't this the same as saying `Result<Option<T>, T::Err>`?

Reply:
Yeah, very similar and equivalent. See #324 for the initial discussion with both of those options proposed. Advantage of this is a single ? in the happy/normal path

Reply:
Yes, what @jbr said :)
I went with this approach because unwrapping twice can be a bit tedious and IMHO the downside of introducing the ParamError enum is lower that the benefit of dealing with only one "effect".